src/EditorFeatures/Core/Implementation/AddImports/AddImportsPasteCommandHandler.cs

```diff
@@ -0,0 +1,120 @@
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.AddImportsPaste)]
+    // Order is important here, this command needs to execute before PasteTracking
+    // since it may modify the pasted span. Paste tracking dismisses if 
+    // the span is modified. It doesn't need to be before FormatDocument, but
+    // this helps the order of execution be more constant in case there 
+    // are problems that arise. This command will always execute the next
+    // command before doing operations.
+    [Order(After = PredefinedCommandHandlerNames.PasteTrackingPaste)]
+    [Order(Before = PredefinedCommandHandlerNames.FormatDocument)]
+    internal class AddImportsPasteCommandHandler : IChainedCommandHandler<PasteCommandArgs>
+    {
+        public string DisplayName => EditorFeaturesResources.Add_Missing_Imports_On_Paste;
+
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AddImportsPasteCommandHandler(IThreadingContext threadingContext)
+            => _threadingContext = threadingContext;
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Capture the pre-paste caret position
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretPosition.HasValue)
+            {
+                return;
+            }
+
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            // Don't perform work if we're inside the interactive window
+            if (args.TextView.IsBufferInInteractiveWindow(args.SubjectBuffer))
+            {
+                return;
+            }
+
+            if (!args.SubjectBuffer.GetFeatureOnOffOption(FeatureOnOffOptions.AddImportsOnPaste))
+            {
+                return;
+            }
+
+            // Create a tracking span from the pre-paste caret position that will grow as text is inserted.
+            var trackingSpan = caretPosition.Value.Snapshot.CreateTrackingSpan(caretPosition.Value.Position, 0, SpanTrackingMode.EdgeInclusive);
+
+            // Applying the post-paste snapshot to the tracking span gives us the span of pasted text.
+            var snapshotSpan = trackingSpan.GetSpan(args.SubjectBuffer.CurrentSnapshot);
+            var textSpan = TextSpan.FromBounds(snapshotSpan.Start, snapshotSpan.End);
+
+            AddMissingImportsForPaste(args, executionContext, textSpan);
+        }
+
+        public CommandState GetCommandState(PasteCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
+
+        private void AddMissingImportsForPaste(PasteCommandArgs args, CommandExecutionContext executionContext, TextSpan textSpan)
```

Comment:
i don't totally see the beneift of extracting this method.  why not just have it be in Execute?

Reply:
It helps me mentally break apart the goal of execute, which is to do/not do `something`. That `something` is this method call. I don't feel strongly if it's blocker, but it does add value to my understanding of the code when I reread it.

Reply:
sorry, github doesn't show me these items.  They're in a collapsed block, and they don't appear when i'm in the 'files' portoin of the review (where i do 100% of my reviewing).

In this case, the major issue i see is that the split doesn't actually make sense.  I agree that the split should be `which is to do/not do something.`.  However, that's not the split here.  For example, despite calling `AddMissingImportsForPaste`, there are still *several* more checks that are done which cause the missing imports to not be added.  If you change it so that all the bail out checks are done frst, and then extract the code to actually add the imports, that would make a lot more sense to me.  Thanks!